File: stock_platform/java/backend/java/src/main/java/com/stockplatform/service/TradeService.java
package com.stockplatform.service;

import com.my.mybatis.flex.core.condition.QueryCondition;
import com.stockplatform.dto.TradeCreateDTO;
import com.stockplatform.dto.TradeResponseDTO;
import com.stockplatform.entity.Position;
import com.stockplatform.entity.Trade;
import com.stockplatform.mapper.PositionMapper;
import com.stockplatform.mapper.TradeMapper;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class TradeService {

    private final TradeMapper tradeMapper;
    private final PositionMapper positionMapper;

    public TradeService(TradeMapper tradeMapper, PositionMapper positionMapper) {
        this.tradeMapper = tradeMapper;
        this.positionMapper = positionMapper;
    }

    @Transactional
    public TradeResponseDTO createTrade(TradeCreateDTO dto) {
        Trade trade = new Trade();
        trade.setAccountId(dto.getAccountId());
        trade.setStockCode(dto.getStockCode());
        trade.setStockName(dto.getStockName());
        trade.setTradeType(dto.getTradeType());
        trade.setQuantity(dto.getQuantity());
        trade.setPrice(dto.getPrice());
        trade.setCommission(dto.getCommission() != null ? dto.getCommission() : 0.0);
        trade.setTradeDate(dto.getTradeDate());
        trade.setCreatedAt(LocalDateTime.now());
        tradeMapper.insertSelective(trade);

        QueryCondition positionCondition = QueryCondition.create()
                .where("account_id = ?", dto.getAccountId())
                .where("stock_code = ?", dto.getStockCode());
        Position position = positionMapper.selectOneByCondition(positionCondition);

        if ("buy".equalsIgnoreCase(dto.getTradeType())) {
            if (position != null) {
                int totalQuantity = position.getQuantity() + dto.getQuantity();
                double totalCost = position.getQuantity() * position.getAvgCost()
                        + dto.getQuantity() * dto.getPrice();
                position.setAvgCost(totalCost / totalQuantity);
                position.setQuantity(totalQuantity);
                position.setUpdatedAt(LocalDateTime.now());
                positionMapper.update(position);
            } else {
                position = new Position();
                position.setAccountId(dto.getAccountId());
                position.setStockCode(dto.getStockCode());
                position.setStockName(dto.getStockName());
                position.setQuantity(dto.getQuantity());
                position.setAvgCost(dto.getPrice());
                position.setCreatedAt(LocalDateTime.now());
                position.setUpdatedAt(LocalDateTime.now());
                positionMapper.insertSelective(position);
            }
        } else if ("sell".equalsIgnoreCase(dto.getTradeType())) {
            if (position != null) {
                double profit = (dto.getPrice() - position.getAvgCost()) * dto.getQuantity()
                        - (dto.getCommission() != null ? dto.getCommission() : 0.0);
                trade.setProfit(profit);
                tradeMapper.update(trade);

                position.setQuantity(position.getQuantity() - dto.getQuantity());
                position.setUpdatedAt(LocalDateTime.now());
                if (position.getQuantity() <= 0) {
                    positionMapper.deleteById(position.getId());
                } else {
                    positionMapper.update(position);
                }
            }
        }

        return TradeResponseDTO.fromEntity(trade);
    }

    public List<TradeResponseDTO> listTrades(Integer accountId, LocalDateTime startDate, LocalDateTime endDate) {
        QueryCondition condition = QueryCondition.create();
        if (accountId != null) {
            condition.where("account_id = ?", accountId);
        }
        if (startDate != null) {
            condition.where("trade_date >= ?", startDate);
        }
        if (endDate != null) {
            condition.where("trade_date <= ?", endDate);
        }
        condition.orderBy("trade_date", false);

        return tradeMapper.selectAllByCondition(condition).stream()
                .map(TradeResponseDTO::fromEntity)
                .collect(Collectors.toList());
    }

    public long getTradeCount() {
        return tradeMapper.selectCount(null);
    }
}

File: stock_platform/python/backend/python/services/trade_service.py
from models import Trade, Position
from schemas import TradeCreate
from datetime import datetime
from typing import Optional


async def create_trade(trade: TradeCreate) -> Trade:
    db_trade = await Trade.create(
        account_id=trade.account_id,
        stock_code=trade.stock_code,
        stock_name=trade.stock_name,
        trade_type=trade.trade_type,
        quantity=trade.quantity,
        price=trade.price,
        commission=trade.commission,
        trade_date=trade.trade_date
    )

    position = await Position.filter(
        account_id=trade.account_id,
        stock_code=trade.stock_code
    ).first()

    if trade.trade_type == "buy":
        if position:
            total_quantity = position.quantity + trade.quantity
            total_cost = position.quantity * position.avg_cost + trade.quantity * trade.price
            position.avg_cost = total_cost / total_quantity
            position.quantity = total_quantity
            position.updated_at = datetime.now()
            await position.save()
        else:
            await Position.create(
                account_id=trade.account_id,
                stock_code=trade.stock_code,
                stock_name=trade.stock_name,
                quantity=trade.quantity,
                avg_cost=trade.price
            )
    elif trade.trade_type == "sell":
        if position:
            profit = (trade.price - position.avg_cost) * trade.quantity - trade.commission
            db_trade.profit = profit
            await db_trade.save()
            position.quantity -= trade.quantity
            position.updated_at = datetime.now()
            if position.quantity <= 0:
                await position.delete()
            else:
                await position.save()

    return db_trade


async def list_trades(
    account_id: int | None = None,
    start_date: datetime | None = None,
    end_date: datetime | None = None
) -> list[Trade]:
    query = Trade.all()
    if account_id is not None:
        query = query.filter(account_id=account_id)
    if start_date:
        query = query.filter(trade_date__gte=start_date)
    if end_date:
        query = query.filter(trade_date__lte=end_date)
    return await query.order_by("-trade_date")